File: src/grid.lua
---@alias Cell { alive: boolean }
---@alias CellAlpha { [number]: { [number]: number } }

---@class Grid
---@field grid Cell[][]
---@field next_grid Cell[][]
---@field rows number
---@field cols number
local Grid = {}
Grid.__index = Grid

---@param rows number
---@param cols number
---@return Grid, CellAlpha
function Grid.new(rows, cols)
  local self = setmetatable({
    grid = {},
    next_grid = {},
    rows = rows,
    cols = cols
  }, Grid)

  -- Initialize the grid and next_grid
  local cell_alpha = {}
  for i = 1, rows do
    self.grid[i] = {}
    self.next_grid[i] = {}
    cell_alpha[i] = {}
    for j = 1, cols do
      self.grid[i][j] = { alive = false }
      self.next_grid[i][j] = { alive = false }
      cell_alpha[i][j] = 0
    end
  end

  return self, cell_alpha
end

---@param cell_alpha CellAlpha
function Grid:clear(cell_alpha)
  for i = 1, self.rows do
    for j = 1, self.cols do
      self.grid[i][j].alive = false
      cell_alpha[i][j] = 0
    end
  end
end

---@param cell_alpha CellAlpha
function Grid:randomize(cell_alpha)
  for i = 1, self.rows do
    for j = 1, self.cols do
      self.grid[i][j].alive = math.random() > 0.85
      cell_alpha[i][j] = self.grid[i][j].alive and 1 or 0
    end
  end
end

---@param cell_alpha CellAlpha
---@param row number
---@param col number
function Grid:toggle_cell(cell_alpha, row, col)
  if row >= 1 and row <= self.rows and col >= 1 and col <= self.cols then
    self.grid[row][col].alive = not self.grid[row][col].alive
    cell_alpha[row][col] = self.grid[row][col].alive and 1 or 0
  end
end

---@param row number
---@param col number
---@return number
function Grid:count_cell_neighbors_at(row, col)
  local count = 0
  for i = -1, 1 do
    for j = -1, 1 do
      -- Skip the cell itself
      if i == 0 and j == 0 then goto continue end

      -- Check the bounds of the grid
      local r = row + i
      local c = col + j
      if r >= 1 and r <= self.rows and c >= 1 and c <= self.cols and self.grid[r][c].alive then
        count = count + 1
      end

      ::continue::
    end
  end

  return count
end

---@param cell_alpha CellAlpha
function Grid:next_generation(cell_alpha)
  -- Prepare the next generation grid
  for i = 1, self.rows do
    for j = 1, self.cols do
      local neighbors = self:count_cell_neighbors_at(i, j)
      self.next_grid[i][j].alive = neighbors == 3 or (self.grid[i][j].alive and neighbors == 2)
    end
  end

  -- Update the current grid and cell_alpha
  for i = 1, self.rows do
    for j = 1, self.cols do
      self.grid[i][j].alive = self.next_grid[i][j].alive
      cell_alpha[i][j] = self.grid[i][j].alive and 1 or 0
    end
  end
end

return Grid

File: src/grid.py
import random


class Cell:
	def __init__(self, state: bool = False):
		self.state = state

	def is_alive(self) -> bool:
		return self.state

	def set_state(self, state: bool) -> None:
		self.state = state

	def toggle(self) -> None:
		self.state = not self.state


class Grid:
	def __init__(self, width: int, height: int):
		self.width = width
		self.height = height
		self.cells = [[Cell() for _ in range(width)] for _ in range(height)]

	def get_cell(self, x: int, y: int) -> bool:
		if 0 <= x < self.width and 0 <= y < self.height:
			return self.cells[y][x].is_alive()
		return False

	def set_cell(self, x: int, y: int, state: bool) -> None:
		if 0 <= x < self.width and 0 <= y < self.height:
			self.cells[y][x].set_state(state)

	def toggle_cell(self, x: int, y: int) -> None:
		if 0 <= x < self.width and 0 <= y < self.height:
			self.cells[y][x].toggle()

	def count_neighbors(self, x: int, y: int) -> int:
		count = 0
		for dx in [-1, 0, 1]:
			for dy in [-1, 0, 1]:
				if dx == 0 and dy == 0:
					continue
				nx, ny = x + dx, y + dy
				if self.get_cell(nx, ny):
					count += 1
		return count

	def update(self) -> None:
		new_cells = [[Cell() for _ in range(self.width)] for _ in range(self.height)]

		for y in range(self.height):
			for x in range(self.width):
				live_neighbors = self.count_neighbors(x, y)
				current_state = self.get_cell(x, y)

				if current_state:
					new_cells[y][x].set_state(live_neighbors in [2, 3])
				else:
					new_cells[y][x].set_state(live_neighbors == 3)

		self.cells = new_cells

	def randomize(self) -> None:
		for y in range(self.height):
			for x in range(self.width):
				self.cells[y][x].set_state(random.random() < 0.3)

	def clear(self) -> None:
		self.cells = [[Cell() for _ in range(self.width)] for _ in range(self.height)]
